lib/radius/dict.ex

```diff
@@ -1,327 +1,155 @@
-        :ets.lookup_element(mod, key, 2)
-      rescue
-        ArgumentError ->
-          raise EntryNotFoundError, type: mod, key: key
+        {acc_vendors, acc_attributes, acc_values}
       end
-    end
-  end
-
-  # module DictEntry
-
-  defmodule Vendor do
-    use DictEntry
-    defstruct id: nil, name: nil, format: {1, 1}, file: "(unknown)", line: 0
-
-    def index_for(val) do
-      [id: val.id, name: val.name]
-    end
-
-    def by_name(nil) do
-      %Vendor{}
-    end
-
-    def by_name(name) do
-      lookup!(name: name)
-    end
-
-    def by_id(id) do
-      lookup!(id: id)
-    end
-  end
-
-  # module Vendor
-
-  defmodule Attribute do
-    use DictEntry
-    defstruct id: nil, name: nil, type: nil, opts: [], vendor: nil, file: "(unknown)", line: 0
-
-    def on_insert(val) do
-      v = Vendor.by_name(val.vendor)
-      %{val | vendor: v}
-    end
-
-    def index_for(val) do
-      [id: {val.vendor.id, val.id}, name: val.name]
-    end
-
-    def by_name(name) do
-      lookup!(name: name)
-    end
-
-    def by_id(vendor \\ nil, id) do
-      lookup!(id: {vendor, id})
-    end
-  end
-
-  # module Attribute
-
-  defmodule Value do
-    use DictEntry
-    defstruct name: nil, attr: nil, value: nil, file: "(unknown)", line: 0
-
-    def on_insert(val) do
-      a = Attribute.lookup!(name: val.attr)
-      %{val | attr: a}
-    end
-
-    def index_for(val) do
-      [
-        value: {val.attr.vendor.id, val.attr.id, val.value},
-        name: {val.attr.vendor.name, val.attr.name, val.name}
-      ]
-    end
+    )
 
-    def by_name(vendor \\ nil, attr, name) do
-      lookup!(name: {vendor, attr, name})
-    end
+  for attribute <- attributes do
+    attr_fun_name = "attr_#{attribute.name}" |> String.replace("-", "_") |> String.to_atom()
 
-    def by_value(vendor \\ nil, attr, name) do
-      lookup!(value: {vendor, attr, name})
-    end
+    defmacro unquote(attr_fun_name)(), do: unquote(attribute.id)
+    defmacro unquote(attr_fun_name)(val), do: {unquote(attribute.id), val}
+    def attribute_by_id(unquote(attribute.id)), do: unquote(Macro.escape(attribute))
+    def attribute_by_name(unquote(attribute.name)), do: unquote(Macro.escape(attribute))
   end
 
-  # module Value
+  for val <- values do
+    val_fun_name = "val_#{val.attr}_#{val.name}" |> String.replace("-", "_") |> String.to_atom()
 
-  defp init_ets() do
-    Vendor.init!()
-    Attribute.init!()
-    Value.init!()
-    :ok
+    defmacro unquote(val_fun_name)(), do: unquote(val.value)
+    def value_by_value(unquote(val.attr), unquote(val.value)), do: unquote(Macro.escape(val))
+    def value_by_name(unquote(val.attr), unquote(val.name)), do: unquote(Macro.escape(val))
   end
 
-  defmodule ParserError do
-    defexception file: "<unknown>", line: -1, msg: nil
-
-    def exception({:error, {line, _, msg}}) do
-      %ParserError{line: line, msg: msg}
-    end
+  for vendor <- vendors do
+    mod = Module.concat(__MODULE__, :"Vendor#{vendor.name}")
```

Comment:
should you also mangle the module name here?

Reply:
Good catch!